doc/man7/openssl-core.h.pod

```diff
@@ -0,0 +1,251 @@
+Arrays of this type are passed on demand from the providers to the
+OpenSSL libraries to describe what algorithms the providers provide
+implementations of, and with what properties.
+Arrays of this type must be terminated with a tuple having function
+identity zero and function pointer C<NULL>.
+
+The algorithm names and property definitions are defined by the
+providers.
+
+=item C<OSSL_PARAM>
+
+This type is a structure that allows passing arbitrary data between
+the OpenSSL libraries and the providers in a form that minimises
+intimate knowledge of each other's internal structures.
+It's normally passed in arrays, where the array is terminated with an
+element where all fields are zero (for non-pointers) or C<NULL> (for
+pointers).
+
+These arrays can be used both to pass parameter data to some function,
+but can also be used to request data from some function.
+For a request, this description talks about a requesting function that
+provides the C<OSSL_PARAM> array, and a responding function that fills
+in the buffers that the C<OSSL_PARAM> elements point at.
+
+Regardless of the use, the array as well as any buffer space it points
+to must be provided by the caller (the requesting function).
+
+The fields of this structure are:
+
+=over 4
+
+=item C<key>
+
+The identity of the parameter in the form of a string.
+
+These identities are defined by the entity that provides these
+parameters, i.e. names for parameters provided by the OpenSSL
+libraries are defined by the libraries, and names for parameters
+provided by providers are defined by those providers, except for the
+pointer form of strings (see data type descriptions below)
+
+=item C<buffer>
+
+=item C<buffer_size>
+
+C<buffer> is a pointer to the memory where the parameter data is or
+shall be stored, and C<buffer_size> is its size in bytes.
+The organization of the data depends on the parameter type and flag.
+
+=item C<return_size>
+
+When an array of C<OSSL_PARAM> is used to request data, the responding
+function must set this field to indicate the actual size of the data
+stored in memory pointed at by the C<buffer> pointer, or the needed
+buffer size if the size indicated by C<buffer_size> is too small.
+
+=item C<data_type>
+
+=for comment It's still debated if this field should be present, or if
+the type should always be implied by how it's used.  Either way, these
+data types will have to be passed together with the names as an array
+of OSSL_ITEM, for discovery purposes.
+
+The C<data_type> is a value that describes the type and organization of
+the data, and can be one of:
+
+=over 4
+
+=item C<OSSL_PARAM_INTEGER>
+
+=item C<OSSL_PARAM_UNSIGNED_INTEGER>
+
+The parameter data is an integer (signed or unsigned) of arbitrary
+length, organized in native form, i.e. most significant byte first on
+Big-Endian systems, and least significant byte first on Little-Endian
+systems.
+
+=item C<OSSL_PARAM_REAL>
+
+=for comment It's still debated if we need this or not.
```

Comment:
Entropy sources?

Reply:
If that's the only reason, just change the entropy value to be a scaled integer.

Reply:
Not sure I understand why entropy sources would be passed as real numbers. They usually come as byte streams, no?

Reply:
The quality estimate of the entropy is usually presented as a real value.  A scaled integer would work, but there is no pain in supporting floating point values now and not needing to support scaled integers **and** floating point values when we need the latter for something else later.

We have a chance to clean the infrastructure up now.  It is unlikely to come around again.

Reply:
Ah, good point.

A question that we will always need to ask ourselves is what is a suitable use of `OSSL_PARAM`...  If entropy quality will always be measured the exact same way, with a double, it might be that a specific callback is more suitable.  Moreover, I see the `OSSL_PARAM` as a means of communication between the application and the provider rather than between the library internals and the provider (libcrypto should mostly just offer functions that will just pass along `OSSL_PARAM` arrays).

And yes, I know, this manual currently tells a different story.  I'm reworking it as we speak.

Reply:
Embedded and IoT systems might not have floating point.  If you can get away without it, please do so.

Reply:
Has this ship flown already?

From the public headers:
> void RAND_add(const void *buf, int num, double randomness);

There are also doubles in `crypto/bio/b_print.c`.

Reply:
Not sure why `BIO` is a factor in this...

Also, *have we decided to support a provider backed RAND API*?  I can't see that we have said this in any clear way anywhere, and if we are, we should do so **soon**

Reply:
I was absolutely of the expectation that the RAND API would be provider backed.

Reply:
Okie.

I think we should make that explicit, 'cause we haven't as far as I know.  Not in the design document or anywhere else.

But, I welcome that addition.

Reply:
We can live with system calls for the _entropy_ source but the DRBGs **_must_** be inside the module.

So, yes the RAND API should be provider backed.